Core/pinot-search.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH PINOT-SEARCH "1" "May 2026" "pinot 1.24" "User Commands"
.SH NAME
pinot-search \- Query search engines from the command-line
.SH SYNOPSIS
.B pinot-search
[\fI\,OPTIONS\/\fR] \fI\,SEARCHENGINETYPE SEARCHENGINENAME|SEARCHENGINEOPTION QUERYINPUT\/\fR
.SH DESCRIPTION
ModuleFactory::loadModules: xapian is supported by libxapianbackend.so
pinot\-search \- Query search engines from the command\-line
.SH OPTIONS
.TP
\fB\-d\fR, \fB\-\-datefirst\fR
sort by date then by relevance
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-l\fR, \fB\-\-locationonly\fR
only show the location of each result
.TP
\fB\-m\fR, \fB\-\-max\fR
maximum number of results (default 10)
.TP
\fB\-r\fR, \fB\-\-storedquery\fR
query input is the name of a stored query
.TP
\fB\-s\fR, \fB\-\-stemming\fR
stemming language (in English)
.TP
\fB\-c\fR, \fB\-\-tocsv\fR
file to export results in CSV format to
.TP
\fB\-x\fR, \fB\-\-toxml\fR
file to export results in XML format to
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.PP
Supported search engine types are: 'opensearch' 'sherlock' 'xapian'
.SH EXAMPLES
pinot\-search opensearch /usr/share/pinot/engines/KrustyDescription.xml "clowns"
.PP
pinot\-search \-\-max 20 sherlock /usr/share/pinot/engines/Bozo.src "clowns"
.PP
pinot\-search xapian ~/.pinot/index "label:Clowns"
.PP
pinot\-search \-\-stemming english xapian somehostname:12345 "clowning"
.SH "REPORTING BUGS"
Report bugs to fabrice.colin@gmail.com
.PP
.br
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/old\-licenses/gpl\-2.0.html>.
.br
There is NO WARRANTY, to the extent permitted by law.


NEWS
2026/05/02 version_1_2_4
General :
 - switched to C++17
Collect :
 - Neon-based downloader was lagging behind

2025/11/09 version_1_2_3
Collect :
 - curl 8.16 type fix
 - the HTML filter supports anchors titles
IndexSearch :
 - updated the Wikipedia plugin
UI :
 - default 64x64 icon

2024/10/27 version_1_2_2
General :
 - const correctness and Debian unstable build fixes
 - support for exiv2 0.28
 - require libarchive >= 3.0.0
 - require libxml++ >= 3.2.0
 - several warning fixes

2022/02/22 version_1_2_1
General :
 - removed code paths around diacritics sensitivity
 - removed code paths pertaining to other query languages
 - auto-generate the ChangeLog file
IndexSearch :
 - removed historical checks on old Xapian releases
 - replaced homegrown abstract solution with Xapian's
UI :
 - fixed query history

2021/10/16 version_1_2_0
General :
 - dropped xdgmime in favour of gio
 - require giomm >= 2.6
 - run rst2txt on RST files, if available
IndexSearch :
 - removed obsolete search plugins, dead code
 - added a plugin for Arxiv
 - better results with the dir query filter
 - fixed size range queries
 - synced with how Xapian Omega indexes file extensions
 - added sort by file size (descending)
 - pinot-index shows document IDs to facilitate troubleshooting
 with xapian-delve
 - require Xapian >= 1.4.10
 - reindexing is recommended.
Daemon :
 - fixed a bug that would cause the loss of documents for one directory
 when crawling ends for the next one
 - overhauled the D-Bus implementation, extended the interface
 - dropped dependency on dbus-glib
 - reduced flushing
 - a IndexFlushEpoch property indicates the last time index changed
 were flushed to disk
 - rely on org.freedesktop.UPower's OnBattery on systems with a battery
 - implemented the org.gnome.Shell.SearchProvider2 interface. The content
 of documents in directories indexed by the daemon can be searched through
 the Gnome 3 Shell.
UI :
 - removed the gtk2 UI, refreshed the gtk3 UI
 - the gtk3 UI is loaded from a Glade XML file
 - reduced direct reading of the daemon's index
 - reopen the daemon's index as per the IndexFlushEpoch property
 - fixed file imports and error reporting
 - stored queries' results may be sorted by file size
 - rationalized activation and behaviour of the Edit menuitems
 - pinot may be started on any new query with -q/--query TERMS
 - require gtkmm >= 3.24

2020/01/09 version_1_1_0
General:
 - updated gmime, textcat, exiv, glib, OpenSSL and Xapian dependencies

2015/06/11 version_1_0_9
Tokenize :
 - new JSON filter
 - streamlined the mbox filter a bit
 - pinot-index looks for filters in the same locations as the UI
IndexSearch :
 - sort documents by date, in ascending or descending order

2014/07/18 version_1_0_8
General :
 - don't install the xdgmime files
 - use libnotify if --enable-libnotify=yes is specified
 - fixed check on library symbols that could cause a crash, added
 support for LLVM-built filters
 - general clean up
SQL :
 - refactored the DB interface

2014/05/22 version_1_0_7
General:
 - compilation fixes
 - prefer off_t for file sizes, offsets, buffer lengths
Monitor :
 - try and reapply watches on directories that have just been removed
Tokenize :
 - refactored encoding conversion
IndexSearch :
 - fixed build with libexttextcat

2013/05/26 version_1_0_6
IndexSearch:
 - support boost 1.50's Spirit
SQL :
 - better handling of potential errors while stepping through results
Collect :
 - minor fixes to curl backend

2013/03/03 version_1_0_5
IndexSearch:
 - fixes to abstract generation, CJKV tokenization

2013/02/11 version_1_0_4
IndexSearch:
 - fix stripping of diacritics
 - stem subject terms

2013/01/14 version_1_0_3
IndexSearch :
 - fix a Unicode handling issue introduced in 1.01
UI :
 - updated French translation by Eliovir

2012/11/04 version_1_0_2
General :
 - turn memory pooling off by default to avoid issues with newer boost
UI :
 - new Czech translation by Zbyněk Schwarz
 - updated Japanese translation by Takafumi Arakaki
 - updated Brazilian Portuguese translation by Adriano Steffler

2012/08/27 version_1_0_1
General :
 - run rst2html on RST files, if available, and if RST files are detected as such.
 See http://code.google.com/p/pinot-search/issues/detail?id=12
Tokenize :
 - better mbox parts extraction
IndexSearch :
 - dropped unac in favour of own code, resulting in faster indexing.
 - pinot-index --override MIMETYPE:EXTENSION overrides MIME type detection
 based on files extensions

2012/06/16 version_1_0_0
General :
 - install headers and libraries
 - prefer default programs that don't support URIs to view local files
Deskbar :
 - dropped support for Deskbar since it's now dead
Tokenize :
 - better mbox parsing. The internal part numbering scheme has changed;
 reindexing email is recommended.
IndexSearch :
 - support for LibreOffice's libexttextcat v3.2
 - fixed the Google plugin
 - renamed the Freshmeat plugin to Freecode
UI :
 - updated Simplified Chinese translation by happymeng
 - updated German translation by Gena Haltmair
 - initial GTK+ 3 port. Enable with "./configure ... --enable-gtkmm3=yes"

2011/11/07 version_0_9_8
Tokenize :
 - new exiv2-based filter
 - new chmlib-based filter
IndexSearch :
 - support for LibreOffice's libexttextcat v3.1, and possibly v3.1.1
 - dropped plugins for Yahoo! REST API, Yahoo! BOSS, Google Code Search
  and RollYO
UI :
 - query results and view history are expired after 6 months
 - updated Dutch translation by Martijn Verstrate and Tico
 - updated German translation by Fitoschido
 - updated Russian translation by Nikolay Kachanov
 - updated Spanish translation by pkramerruiz

2011/01/09 version_0_9_7
General :
 - replaced custom memory pool class with Boost's
 - don't try and map more than 2Gb and use shared mappings
Deskbar :
 - install the module where DeskBar > 2.28 expects it to be
IndexSearch :
 - remove dots at the end of terms that don't look like acronyms
 - index components of acronyms and dot-separated terms on their own
Daemon :

Core/pinot-search.cpp
/*
 *  Copyright 2005-2024 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <iostream>
#include <fstream>
#include <string>
#include <giomm/init.h>
#include <glibmm/init.h>

#include "config.h"
#include "NLS.h"
#include "Languages.h"
#include "MIMEScanner.h"
#include "Url.h"
#include "DownloaderFactory.h"
#include "ModuleFactory.h"
#include "ResultsExporter.h"
#include "WebEngine.h"
#include "PinotSettings.h"

using namespace std;

static struct option g_longOptions[] = {
	{"datefirst", 0, 0, 'd'},
	{"help", 0, 0, 'h'},
	{"locationonly", 0, 0, 'l'},
	{"max", 1, 0, 'm'},
	{"storedquery", 0, 0, 'r'},
	{"stemming", 1, 0, 's'},
	{"tocsv", 1, 0, 'c'},
	{"toxml", 1, 0, 'x'},
	{"version", 0, 0, 'v'},
	{0, 0, 0, 0}
};

static void printHelp(void)
{
	map<ModuleProperties, bool> engines;

	// Help
	ModuleFactory::loadModules(string(LIBDIR) + string("/pinot/backends"));
	ModuleFactory::getSupportedEngines(engines);
	ModuleFactory::unloadModules();
	clog << "pinot-search - Query search engines from the command-line\n\n"
		<< "Usage: pinot-search [OPTIONS] SEARCHENGINETYPE SEARCHENGINENAME|SEARCHENGINEOPTION QUERYINPUT\n\n"
		<< "Options:\n"
		<< "  -d, --datefirst           sort by date then by relevance\n"
		<< "  -h, --help                display this help and exit\n"
		<< "  -l, --locationonly        only show the location of each result\n"
		<< "  -m, --max                 maximum number of results (default 10)\n"
		<< "  -r, --storedquery         query input is the name of a stored query\n"
		<< "  -s, --stemming            stemming language (in English)\n"
		<< "  -c, --tocsv               file to export results in CSV format to\n"
		<< "  -x, --toxml               file to export results in XML format to\n"
		<< "  -v, --version             output version information and exit\n"
		<< "Supported search engine types are:";
	for (map<ModuleProperties, bool>::const_iterator engineIter = engines.begin(); engineIter != engines.end(); ++engineIter)
	{
		clog << " '" << engineIter->first.m_name << "'";
	}
	clog << "\n\nExamples:\n"
		<< "pinot-search opensearch " << PINOT_PREFIX << "/share/pinot/engines/KrustyDescription.xml \"clowns\"\n\n"
		<< "pinot-search --max 20 sherlock " << PINOT_PREFIX << "/share/pinot/engines/Bozo.src \"clowns\"\n\n"
		<< "pinot-search xapian ~/.pinot/index \"label:Clowns\"\n\n"
		<< "pinot-search --stemming english xapian somehostname:12345 \"clowning\"\n\n"
		<< "Report bugs to " << PACKAGE_BUGREPORT << endl;
}

int main(int argc, char **argv)
{
	string engineType, option, csvExport, xmlExport, stemLanguage;
	unsigned int maxResultsCount = 10; 
	int longOptionIndex = 0;
	bool printResults = true;
	bool sortByDate = false;
	bool locationOnly = false;
	bool isStoredQuery = false;

	// Look at the options
	int optionChar = getopt_long(argc, argv, "c:dhlm:rs:vx:", g_longOptions, &longOptionIndex);
	while (optionChar != -1)
	{
		switch (optionChar)
		{
			case 'c':
				if (optarg != NULL)
				{
					csvExport = optarg;
					printResults = false;
				}
				break;
			case 'd':
				sortByDate = true;
				break;
			case 'h':
				printHelp();
				return EXIT_SUCCESS;
			case 'l':
				locationOnly = true;
				break;
			case 'm':
				if (optarg != NULL)
				{
					maxResultsCount = (unsigned int )atoi(optarg);
				}
				break;
			case 'r':
				isStoredQuery = true;
				break;
			case 's':
				if (optarg != NULL)
				{
					stemLanguage = optarg;
				}
				break;
			case 'v':
				clog << "pinot-search - " << PACKAGE_STRING << "\n\n"
					<< "This is free software.  You may redistribute copies of it under the terms of\n"
					<< "the GNU General Public License <http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>.\n"
					<< "There is NO WARRANTY, to the extent permitted by law." << endl;
				return EXIT_SUCCESS;
			case 'x':
				if (optarg != NULL)
				{
					xmlExport = optarg;
					printResults = false;
				}
				break;
			default:
				return EXIT_FAILURE;
		}

		// Next option
		optionChar = getopt_long(argc, argv, "c:dhlm:rs:vx:", g_longOptions, &longOptionIndex);
	}

#if defined(ENABLE_NLS)
	bindtextdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);
#endif //ENABLE_NLS

	Glib::init();
	Gio::init();

	if (argc == 1)
	{
		printHelp();
		return EXIT_SUCCESS;
	}

	if ((argc < 4) ||
		(argc - optind != 3))
	{
		clog << "Wrong number of parameters" << endl;
		return EXIT_FAILURE;
	}

	// This will create the necessary directories on the first run
	PinotSettings &settings = PinotSettings::getInstance();
	string confDirectory(PinotSettings::getConfigurationDirectory());

	if (MIMEScanner::initialize(PinotSettings::getHomeDirectory() + "/.local",
		string(SHARED_MIME_INFO_PREFIX)) == false)
	{
		clog << "Couldn't load MIME settings" << endl;
	}
	DownloaderInterface::initialize();
	ModuleFactory::loadModules(string(LIBDIR) + string("/pinot/backends"));
	ModuleFactory::loadModules(confDirectory + "/backends");

	// Localize language names
	Languages::setIntlName(0, _("Unknown"));
	Languages::setIntlName(1, _("Danish"));
	Languages::setIntlName(2, _("Dutch"));
	Languages::setIntlName(3, _("English"));
	Languages::setIntlName(4, _("Finnish"));
	Languages::setIntlName(5, _("French"));
	Languages::setIntlName(6, _("German"));
	Languages::setIntlName(7, _("Hungarian"));
	Languages::setIntlName(8, _("Italian"));
	Languages::setIntlName(9, _("Norwegian"));
	Languages::setIntlName(10, _("Portuguese"));
	Languages::setIntlName(11, _("Romanian"));
	Languages::setIntlName(12, _("Russian"));
	Languages::setIntlName(13, _("Spanish"));
	Languages::setIntlName(14, _("Swedish"));
	Languages::setIntlName(15, _("Turkish"));

	// Load the settings
	settings.load(PinotSettings::LOAD_ALL);

	engineType = argv[optind];
	option = argv[optind + 1];
	char *pQueryInput = argv[optind + 2];

	// Set the query
	QueryProperties queryProps("pinot-search", "");
	if (isStoredQuery == true)
	{
		const map<string, QueryProperties> &queries = settings.getQueries();
		map<string, QueryProperties>::const_iterator queryIter = queries.find(pQueryInput);
		if (queryIter != queries.end())
		{
			queryProps = queryIter->second;
		}
		else
		{
			clog << "Couldn't find stored query " << pQueryInput << endl;

			DownloaderInterface::shutdown();
			MIMEScanner::shutdown();

			return EXIT_FAILURE;
		}
	}
	else
	{
		queryProps.setFreeQuery(pQueryInput);
	}
	queryProps.setStemmingLanguage(stemLanguage);
	queryProps.setMaximumResultsCount(maxResultsCount);
	if (sortByDate == true)
	{
		queryProps.setSortOrder(QueryProperties::DATE_DESC);
	}

	// Which SearchEngine ?
	SearchEngineInterface *pEngine = ModuleFactory::getSearchEngine(engineType, option);
	if (pEngine == NULL)
	{
		clog << "Couldn't obtain search engine instance" << endl;

		DownloaderInterface::shutdown();
		MIMEScanner::shutdown();

		return EXIT_FAILURE;
	}

	// Set up the proxy
	WebEngine *pWebEngine = dynamic_cast<WebEngine *>(pEngine);
	if (pWebEngine != NULL)
	{
		DownloaderInterface *pDownloader = pWebEngine->getDownloader();
		if ((pDownloader != NULL) &&
			(settings.m_proxyEnabled == true) &&
			(settings.m_proxyAddress.empty() == false))
		{
			char portStr[64];

			pDownloader->setSetting("proxyaddress", settings.m_proxyAddress);
			snprintf(portStr, 64, "%u", settings.m_proxyPort);
			pDownloader->setSetting("proxyport", portStr);
			pDownloader->setSetting("proxytype", settings.m_proxyType);
		}

		pWebEngine->setEditableValues(settings.m_editablePluginValues);
	}

	pEngine->setDefaultOperator(SearchEngineInterface::DEFAULT_OP_AND);
	if (pEngine->runQuery(queryProps) == true)
	{
		string resultsPage;
		unsigned int estimatedResultsCount = pEngine->getResultsCountEstimate();

		const vector<DocumentInfo> &resultsList = pEngine->getResults();
		if (resultsList.empty() == false)
		{
			if (printResults == true)
			{
				unsigned int count = 0;

				if (locationOnly == false)
				{
					clog << "Showing " << resultsList.size() << " results of about " << estimatedResultsCount << endl;
				}

				vector<DocumentInfo>::const_iterator resultIter = resultsList.begin();
				while (resultIter != resultsList.end())
				{
					string rawUrl(resultIter->getLocation(true));

					if (locationOnly == false)
					{
						clog << count << " Location : '" << rawUrl << "'"<< endl;
						clog << count << " Title    : " << resultIter->getTitle() << endl;
						clog << count << " Type     : " << resultIter->getType() << endl;
						clog << count << " Language : " << resultIter->getLanguage() << endl;
						clog << count << " Date     : " << resultIter->getTimestamp() << endl;
						clog << count << " Size     : " << resultIter->getSize() << endl;
						clog << count << " Extract  : " << resultIter->getExtract() << endl;
						clog << count << " Score    : " << resultIter->getScore() << endl;
					}
					else
					{
						clog << rawUrl << endl;
					}
					++count;

					// Next
					++resultIter;
				}
			}
			else
			{
				string engineName(ModuleFactory::getSearchEngineName(engineType, option));

				if (csvExport.empty() == false)
				{
					CSVExporter exporter(csvExport, queryProps);

					exporter.exportResults(engineName, maxResultsCount, resultsList);
				}

				if (xmlExport.empty() == false)
				{
					OpenSearchExporter exporter(xmlExport, queryProps);

					exporter.exportResults(engineName, maxResultsCount, resultsList);
				}
			}
		}
		else
		{
			clog << "No results" << endl;
		}
	}
	else
	{
		clog << "Couldn't run query on search engine " << engineType << endl;
	}

	delete pEngine;

	ModuleFactory::unloadModules();
	DownloaderInterface::shutdown();
	MIMEScanner::shutdown();

	return EXIT_SUCCESS;
}
